crud/common/schema.lua

```diff
@@ -0,0 +1,168 @@
+            return nil, err
+        end
+
+        reload_schema_cond:broadcast()
+        reload_in_progress = false
+    end
+
+    return true
+end
+
+-- schema.wrap_func_reload calls func with specified arguments.
+-- func should return `res, err, need_reload`
+-- If function returned error and `need_reload` is true,
+-- then schema is reloaded and one more attempt is performed
+-- (but no more than RELOAD_RETRIES_NUM).
+-- This wrapper is used for functions that can fail if router uses outdated
+-- space schema. In case of such errors these functions returns `need_reload`
+-- for schema-dependent errors.
+function schema.wrap_func_reload(func, ...)
+    local i = 0
+
+    local res, err, need_reload
+    while true do
+        res, err, need_reload = func(...)
+
+        if err == nil or not need_reload then
+            break
+        end
+
+        local ok, reload_schema_err = reload_schema(vshard.router.routeall())
+        if not ok then
+            log.debug("Failed to reload schema: %s", reload_schema_err)
+            break
+        end
+
+        i = i + 1
+        if i > const.RELOAD_RETRIES_NUM then
+            break
+        end
+    end
+
+    return res, err
+end
+
+local function get_space_schema_hash(space)
+    if space == nil then
+        return ''
+    end
+
+    local indexes_info = {}
+    for i = 0, table.maxn(space.index) do
+        local index = space.index[i]
+        if index ~= nil then
+            indexes_info[i] = {
+                unique = index.unique,
+                parts = index.parts,
+                id = index.id,
+                type = index.type,
+                name = index.name,
+                path = index.path,
+            }
+        end
+    end
+
+    local space_info = {
+        format = space:format(),
+        indexes = indexes_info,
+    }
+
+    return digest.murmur(msgpack.encode(space_info))
+end
+
+-- schema.wrap_box_space_func_result pcalls some box.space function
+-- and returns it's result as a table
+-- `{res = ..., err = ..., space_schema_hash = ...}`
+-- space_schema_hash is computed if function failed and
+-- `add_space_schema_hash` is true
+function schema.wrap_box_space_func_result(add_space_schema_hash, space, func_name, ...)
+    local result = {}
+
+    local ok, func_res = pcall(space[func_name], space, ...)
+    if not ok then
+        result.err = func_res
+        if add_space_schema_hash then
+            result.space_schema_hash = get_space_schema_hash(space)
+        end
+    else
+        result.res = func_res
+    end
+
+    return result
+end
+
+-- schema.result_needs_reload checks that schema reload can
+-- be helpful to avoid storage error.
+-- It checks if space_schema_hash returned by storage
+-- is the same as hash of space used on router.
+-- Note, that storage returns `space_schema_hash = nil`
+-- if reloading space format can't avoid the error.
+function schema.result_needs_reload(space, result)
+    if result.space_schema_hash == nil then
```

Comment:
No, `nil` hash means that reloading can't avoid the error.
Empty hash could mean that space doesn't exist on storage (but it can still exist on router). Now this case isn't actual, but in the future when we start to create spaces using crud it could be helpful.